File: samaria/samaria/page/samaria_dashboard/samaria_dashboard.py
# -*- coding: utf-8 -*-
"""
Samaria Executive Dashboard Backend Controller
Frappe Framework Version-16
"""
import frappe
from frappe import _
from frappe.utils import flt, nowdate, add_days


@frappe.whitelist()
def get_dashboard_data(customer=None, from_date=None, to_date=None):
	"""
	Aggregates multi-division metrics across Aggregate, Cement, Medical, and Commercial agreements.
	"""
	data = {
		"metrics": {
			"aggregate": {
				"total_dispatches": 0,
				"delivered_volume": 0.0,
				"shortage_volume": 0.0,
				"net_profit": 0.0,
				"gross_truck_fee": 0.0
			},
			"cement": {
				"total_liftings": 0,
				"total_weight_tons": 0.0,
				"total_shortage_tons": 0.0,
				"total_penalties": 0.0
			},
			"medical": {
				"active_batches": 0,
				"quarantined_batches": 0,
				"near_expiry_batches": 0,
				"pending_requests": 0
			},
			"commercial": {
				"active_agreements": 0,
				"total_contract_value": 0.0
			}
		},
		"charts": {
			"aggregate_trend": {
				"labels": [],
				"datasets": [{"name": _("Delivered Volume (m³)"), "values": []}]
			},
			"cement_by_factory": {
				"labels": [],
				"datasets": [{"values": []}]
			},
			"division_mix": {
				"labels": [_("Aggregate"), _("Cement"), _("Medical"), _("Commercial")],
				"datasets": [{"values": [0, 0, 0, 0]}]
			}
		},
		"recent_dispatches": [],
		"recent_liftings": [],
		"urgent_medical": []
	}

	# 1. Aggregate Deliveries
	if frappe.db.table_exists("Aggregate Delivery"):
		conds = ["docstatus < 2"]
		vals = {}
		if customer:
			conds.append("customer = %(customer)s")
			vals["customer"] = customer
		if from_date:
			conds.append("dispatch_date >= %(from_date)s")
			vals["from_date"] = from_date
		if to_date:
			conds.append("dispatch_date <= %(to_date)s")
			vals["to_date"] = to_date

		where_clause = " AND ".join(conds)

		agg_summary = frappe.db.sql(f"""
			SELECT
				COUNT(name) as total_count,
				COALESCE(SUM(delivered_volume), 0) as del_vol,
				COALESCE(SUM(shortage_volume), 0) as short_vol,
				COALESCE(SUM(net_profit_amount), 0) as profit,
				COALESCE(SUM(gross_truck_fee), 0) as truck_fee
			FROM `tabAggregate Delivery`
			WHERE {where_clause}
		""", vals, as_dict=True)

		if agg_summary:
			row = agg_summary[0]
			data["metrics"]["aggregate"]["total_dispatches"] = row.total_count or 0
			data["metrics"]["aggregate"]["delivered_volume"] = round(flt(row.del_vol), 1)
			data["metrics"]["aggregate"]["shortage_volume"] = round(flt(row.short_vol), 1)
			data["metrics"]["aggregate"]["net_profit"] = round(flt(row.profit), 2)
			data["metrics"]["aggregate"]["gross_truck_fee"] = round(flt(row.truck_fee), 2)

		# Aggregate Monthly/Daily Trend
		trend_rows = frappe.db.sql(f"""
			SELECT dispatch_date, SUM(delivered_volume) as vol
			FROM `tabAggregate Delivery`
			WHERE {where_clause}
			GROUP BY dispatch_date
			ORDER BY dispatch_date DESC
			LIMIT 7
		""", vals, as_dict=True)

		trend_rows.reverse()
		for r in trend_rows:
			data["charts"]["aggregate_trend"]["labels"].append(str(r.dispatch_date))
			data["charts"]["aggregate_trend"]["datasets"][0]["values"].append(round(flt(r.vol), 1))

		# Recent 5 dispatches
		data["recent_dispatches"] = frappe.db.sql(f"""
			SELECT name, pad_number, customer_name, item_name, delivered_volume, net_profit_amount, status, dispatch_date
			FROM `tabAggregate Delivery`
			WHERE {where_clause}
			ORDER BY dispatch_date DESC, creation DESC
			LIMIT 5
		""", vals, as_dict=True)

	# 2. Cement Lifting
	if frappe.db.table_exists("Cement Lifting"):
		c_conds = ["docstatus < 2"]
		c_vals = {}
		if customer:
			c_conds.append("customer = %(customer)s")
			c_vals["customer"] = customer
		if from_date:
			c_conds.append("lifting_date >= %(from_date)s")
			c_vals["from_date"] = from_date
		if to_date:
			c_conds.append("lifting_date <= %(to_date)s")
			c_vals["to_date"] = to_date

		c_where = " AND ".join(c_conds)

		cement_summary = frappe.db.sql(f"""
			SELECT
				COUNT(name) as total_count,
				COALESCE(SUM(buyer_weighbridge_qty), 0) as total_tons,
				COALESCE(SUM(shortage_qty), 0) as short_tons,
				COALESCE(SUM(shortage_penalty), 0) as penalties
			FROM `tabCement Lifting`
			WHERE {c_where}
		""", c_vals, as_dict=True)

		if cement_summary:
			c_row = cement_summary[0]
			data["metrics"]["cement"]["total_liftings"] = c_row.total_count or 0
			data["metrics"]["cement"]["total_weight_tons"] = round(flt(c_row.total_tons), 1)
			data["metrics"]["cement"]["total_shortage_tons"] = round(flt(c_row.short_tons), 1)
			data["metrics"]["cement"]["total_penalties"] = round(flt(c_row.penalties), 2)

		# Cement by Factory chart
		factory_rows = frappe.db.sql(f"""
			SELECT factory, SUM(buyer_weighbridge_qty) as total_tons
			FROM `tabCement Lifting`
			WHERE {c_where} AND factory IS NOT NULL AND factory != ''
			GROUP BY factory
			ORDER BY total_tons DESC
			LIMIT 5
		""", c_vals, as_dict=True)

		for fr in factory_rows:
			data["charts"]["cement_by_factory"]["labels"].append(fr.factory)
			data["charts"]["cement_by_factory"]["datasets"][0]["values"].append(round(flt(fr.total_tons), 1))

		# Recent 5 liftings
		data["recent_liftings"] = frappe.db.sql(f"""
			SELECT name, factory, customer_name, buyer_weighbridge_qty, shortage_qty, status, lifting_date
			FROM `tabCement Lifting`
			WHERE {c_where}
			ORDER BY lifting_date DESC, creation DESC
			LIMIT 5
		""", c_vals, as_dict=True)

	# 3. Medical Batches & Requests
	if frappe.db.table_exists("Medical Batch"):
		data["metrics"]["medical"]["active_batches"] = frappe.db.count("Medical Batch", {"is_quarantined": 0})
		data["metrics"]["medical"]["quarantined_batches"] = frappe.db.count("Medical Batch", {"is_quarantined": 1})

		ninety_days = add_days(nowdate(), 90)
		data["metrics"]["medical"]["near_expiry_batches"] = frappe.db.count(
			"Medical Batch",
			{"expiry_date": ["between", [nowdate(), ninety_days]]}
		)

	if frappe.db.table_exists("Medical Request"):
		data["metrics"]["medical"]["pending_requests"] = frappe.db.count(
			"Medical Request",
			{"status": ["in", ["Submitted", "Quoted", "Approved"]]}
		)
		data["urgent_medical"] = frappe.get_all(
			"Medical Request",
			filters={"status": ["in", ["Submitted", "Quoted", "Approved"]]},
			fields=["name", "customer_name", "priority", "status", "request_date"],
			order_by="creation desc",
			limit=5
		)

	# 4. Sales Agreements
	if frappe.db.table_exists("Sales Agreement"):
		data["metrics"]["commercial"]["active_agreements"] = frappe.db.count("Sales Agreement", {"status": "Active"})
		agr_val = frappe.db.sql("""
			SELECT COALESCE(SUM(total_amount), 0) as total_val
			FROM `tabSales Agreement`
			WHERE status = 'Active'
		""", as_dict=True)
		if agr_val:
			data["metrics"]["commercial"]["total_contract_value"] = round(flt(agr_val[0].total_val), 2)

	# Division mix chart
	data["charts"]["division_mix"]["datasets"][0]["values"] = [
		data["metrics"]["aggregate"]["total_dispatches"],
		data["metrics"]["cement"]["total_liftings"],
		data["metrics"]["medical"]["active_batches"],
		data["metrics"]["commercial"]["active_agreements"]
	]

	return data


@frappe.whitelist()
def get_project_filters():
	"""Fetches list of customers and offloading sites for dashboard filter dropdown."""
	customers = []
	try:
		if frappe.db.table_exists("Customer"):
			customers = frappe.get_all("Customer", fields=["name", "customer_name"], order_by="customer_name asc", limit=50)
	except Exception:
		pass
	return customers

File: samaria/samaria/page/samaria_dashboard/samaria_dashboard.js
/**
 * Samaria Executive Dashboard Page Controller
 * Frappe Framework Version-16
 */

frappe.pages['samaria_dashboard'].on_page_load = function(wrapper) {
	var page = frappe.ui.make_app_page({
		parent: wrapper,
		title: __('Samaria Executive Dashboard'),
		single_column: true
	});

	frappe.breadcrumbs.add("Samaria");
	new SamariaDashboard(page);
};

class SamariaDashboard {
	constructor(page) {
		this.page = page;
		this.filters = {
			customer: null,
			from_date: null,
			to_date: null
		};
		this.init();
	}

	init() {
		this.setup_actions();
		this.setup_filter_fields();
		this.render_layout();
		this.refresh();
	}

	setup_actions() {
		// Quick Action Buttons
		this.page.set_primary_action(
			__('New Dispatch'),
			() => frappe.new_doc('Aggregate Delivery'),
			'add'
		);

		this.page.add_secondary_action(
			__('New Cement Lifting'),
			() => frappe.new_doc('Cement Lifting')
		);

		this.page.add_inner_button(__('New Sales Agreement'), () => {
			frappe.new_doc('Sales Agreement');
		}, __('Quick Actions'));

		this.page.add_inner_button(__('Open Project Reports'), () => {
			frappe.set_route('query-report', 'Aggregate Project Dispatch Report');
		}, __('Reports'));

		this.page.add_inner_button(__('Cement Lifting Report'), () => {
			frappe.set_route('query-report', 'Cement Project Lifting and Balance Report');
		}, __('Reports'));

		this.page.add_inner_button(__('Medical Inventory Report'), () => {
			frappe.set_route('query-report', 'Medical Project Inventory Report');
		}, __('Reports'));

		this.page.add_inner_button(__('Financial Summary Report'), () => {
			frappe.set_route('query-report', 'Project Financial Summary Report');
		}, __('Reports'));
	}

	setup_filter_fields() {
		// Customer / Project Filter
		this.customer_field = this.page.add_field({
			fieldname: 'customer',
			label: __('Project / Customer'),
			fieldtype: 'Link',
			options: 'Customer',
			change: () => {
				this.filters.customer = this.customer_field.get_value();
				this.refresh();
			}
		});

		// From Date Filter
		this.from_date_field = this.page.add_field({
			fieldname: 'from_date',
			label: __('From Date'),
			fieldtype: 'Date',
			change: () => {
				this.filters.from_date = this.from_date_field.get_value();
				this.refresh();
			}
		});

		// To Date Filter
		this.to_date_field = this.page.add_field({
			fieldname: 'to_date',
			label: __('To Date'),
			fieldtype: 'Date',
			change: () => {
				this.filters.to_date = this.to_date_field.get_value();
				this.refresh();
			}
		});
	}

	render_layout() {
		this.$container = $(`
			<div class="samaria-dashboard-container">
				<!-- KPI Cards Row -->
				<div class="samaria-kpi-grid" id="samaria-kpis">
					<div class="text-muted">${__('Loading KPI metrics...')}</div>
				</div>

				<!-- Charts Row -->
				<div class="samaria-charts-grid">
					<div class="samaria-chart-card">
						<h4>${__('Aggregate Dispatch Volume Trend (m³)')}</h4>
						<div id="chart-aggregate-trend" style="min-height: 240px;"></div>
					</div>
					<div class="samaria-chart-card">
						<h4>${__('Cement Lifting Distribution by Factory')}</h4>
						<div id="chart-cement-factory" style="min-height: 240px;"></div>
					</div>
				</div>

				<!-- Tables Row -->
				<div class="samaria-tables-grid">
					<div class="samaria-table-card">
						<h4>
							<span>${__('Recent Aggregate Deliveries')}</span>
							<a href="/app/aggregate-delivery" class="text-muted" style="font-size: 0.8rem;">${__('View All')} &rarr;</a>
						</h4>
						<div id="recent-aggregate-table"></div>
					</div>
					<div class="samaria-table-card">
						<h4>
							<span>${__('Recent Cement Liftings')}</span>
							<a href="/app/cement-lifting" class="text-muted" style="font-size: 0.8rem;">${__('View All')} &rarr;</a>
						</h4>
						<div id="recent-cement-table"></div>
					</div>
				</div>
			</div>
		`).appendTo(this.page.main);
	}

	refresh() {
		frappe.call({
			method: 'samaria.samaria.page.samaria_dashboard.samaria_dashboard.get_dashboard_data',
			args: {
				customer: this.filters.customer,
				from_date: this.filters.from_date,
				to_date: this.filters.to_date
			},
			callback: (r) => {
				if (r.message) {
					this.render_kpis(r.message.metrics);
					this.render_charts(r.message.charts);
					this.render_tables(r.message);
				}
			}
		});
	}

	render_kpis(metrics) {
		const m = metrics;
		const html = `
			<div class="samaria-kpi-card orange">
				<div class="samaria-kpi-title">
					<span>${__('Aggregate Volume')}</span>
					<span class="indicator orange">${m.aggregate.total_dispatches} ${__('Loads')}</span>
				</div>
				<div class="samaria-kpi-value">${m.aggregate.delivered_volume} <span style="font-size: 1rem; font-weight: 500;">m³</span></div>
				<div class="samaria-kpi-subtext">
					<span>${__('Net Margin:')} <strong>ETB ${format_number(m.aggregate.net_profit, null, 2)}</strong></span>
				</div>
			</div>

			<div class="samaria-kpi-card emerald">
				<div class="samaria-kpi-title">
					<span>${__('Cement Lifted')}</span>
					<span class="indicator green">${m.cement.total_liftings} ${__('Tickets')}</span>
				</div>
				<div class="samaria-kpi-value">${m.cement.total_weight_tons} <span style="font-size: 1rem; font-weight: 500;">Tons</span></div>
				<div class="samaria-kpi-subtext">
					<span>${__('Shortage Penalty:')} <strong>ETB ${format_number(m.cement.total_penalties, null, 2)}</strong></span>
				</div>
			</div>

			<div class="samaria-kpi-card rose">
				<div class="samaria-kpi-title">
					<span>${__('Medical Batches')}</span>
					<span class="indicator red">${m.medical.quarantined_batches} ${__('Quarantine')}</span>
				</div>
				<div class="samaria-kpi-value">${m.medical.active_batches} <span style="font-size: 1rem; font-weight: 500;">Active</span></div>
				<div class="samaria-kpi-subtext">
					<span>${__('Expiring <90 Days:')} <strong>${m.medical.near_expiry_batches}</strong> | ${__('Pending Req:')} <strong>${m.medical.pending_requests}</strong></span>
				</div>
			</div>

			<div class="samaria-kpi-card purple">
				<div class="samaria-kpi-title">
					<span>${__('Active Contracts')}</span>
					<span class="indicator purple">${m.commercial.active_agreements} ${__('Active')}</span>
				</div>
				<div class="samaria-kpi-value"><span style="font-size: 1.1rem; font-weight: 600;">ETB</span> ${format_number(m.commercial.total_contract_value, null, 2)}</div>
				<div class="samaria-kpi-subtext">
					<span>${__('Aggregate, Cement & Medical Agreements')}</span>
				</div>
			</div>
		`;
		this.$container.find('#samaria-kpis').html(html);
	}

	render_charts(charts) {
		// 1. Aggregate Trend Chart
		const trendData = charts.aggregate_trend;
		const $trendWrapper = this.$container.find('#chart-aggregate-trend');
		$trendWrapper.empty();

		if (trendData.labels && trendData.labels.length > 0) {
			new frappe.Chart($trendWrapper[0], {
				data: trendData,
				type: 'line',
				height: 220,
				colors: ['#f97316'],
				lineOptions: { hideDots: 0, regionFill: 1 }
			});
		} else {
			$trendWrapper.html(`<div class="text-muted text-center" style="padding: 4rem 0;">${__('No aggregate dispatch data in selected range')}</div>`);
		}

		// 2. Cement Distribution Chart
		const cementData = charts.cement_by_factory;
		const $cementWrapper = this.$container.find('#chart-cement-factory');
		$cementWrapper.empty();

		if (cementData.labels && cementData.labels.length > 0) {
			new frappe.Chart($cementWrapper[0], {
				data: cementData,
				type: 'donut',
				height: 220,
				colors: ['#10b981', '#06b6d4', '#3b82f6', '#8b5cf6', '#f59e0b']
			});
		} else {
			$cementWrapper.html(`<div class="text-muted text-center" style="padding: 4rem 0;">${__('No cement lifting records found')}</div>`);
		}
	}

	render_tables(data) {
		// Recent Dispatches Table
		const $aggTable = this.$container.find('#recent-aggregate-table');
		if (data.recent_dispatches && data.recent_dispatches.length > 0) {
			let rows = data.recent_dispatches.map(d => `
				<tr>
					<td>
						<a href="/app/aggregate-delivery/${d.name}"><strong>${d.name}</strong></a><br>
						<small class="text-muted">${d.customer_name || ''}</small>
					</td>
					<td>${d.item_name || 'Aggregate'}<br><small class="text-muted">Pad: ${d.pad_number || '-'}</small></td>
					<td><strong>${d.delivered_volume || 0}</strong> m³</td>
					<td>
						<span class="samaria-badge-sm ${d.status === 'Delivered' || d.status === 'Settled' ? 'green' : 'blue'}">
							${d.status}
						</span>
					</td>
				</tr>
			`).join('');

			$aggTable.html(`
				<table class="samaria-dash-table">
					<thead>
						<tr>
							<th>${__('Dispatch / Site')}</th>
							<th>${__('Material')}</th>
							<th>${__('Delivered')}</th>
							<th>${__('Status')}</th>
						</tr>
					</thead>
					<tbody>${rows}</tbody>
				</table>
			`);
		} else {
			$aggTable.html(`<div class="text-muted text-center" style="padding: 2rem 0;">${__('No recent aggregate dispatches')}</div>`);
		}

		// Recent Cement Liftings Table
		const $cementTable = this.$container.find('#recent-cement-table');
		if (data.recent_liftings && data.recent_liftings.length > 0) {
			let rows = data.recent_liftings.map(c => `
				<tr>
					<td>
						<a href="/app/cement-lifting/${c.name}"><strong>${c.name}</strong></a><br>
						<small class="text-muted">${c.customer_name || ''}</small>
					</td>
					<td>${c.factory || '-'}<br><small class="text-muted">${c.lifting_date || ''}</small></td>
					<td><strong>${c.buyer_weighbridge_qty || 0}</strong> Tons</td>
					<td>
						<span class="samaria-badge-sm ${c.status === 'Delivered' || c.status === 'Verified' ? 'green' : 'blue'}">
							${c.status}
						</span>
					</td>
				</tr>
			`).join('');

			$cementTable.html(`
				<table class="samaria-dash-table">
					<thead>
						<tr>
							<th>${__('Lifting Ref / Buyer')}</th>
							<th>${__('Factory / Date')}</th>
							<th>${__('Weight')}</th>
							<th>${__('Status')}</th>
						</tr>
					</thead>
					<tbody>${rows}</tbody>
				</table>
			`);
		} else {
			$cementTable.html(`<div class="text-muted text-center" style="padding: 2rem 0;">${__('No recent cement liftings')}</div>`);
		}
	}
}
